src/bbd/geocoder/geocoder.py

```diff
@@ -0,0 +1,582 @@
+             "County" : str county component of address
+             "State" : str state component of address
+             "Zip5" : str 5-digit postal code for address
+             "Country" : str country component address
+                         (if applicable, default will set country to US)
+        For list of strings:
+            Each item should str full address to pass as query.
+
+    email : str
+        A valid email address as str to pass to Nominatim as user_agent.
+        See Nominatim's Usage Policy.
+
+    path : str or pathlib Path object
+        A path to where geocoded data will be saved.
+
+    batch_size : (Optional) int
+        Number of batches to geocode at a time. Useful for running subset of
+        data. Will determin size of batch for run(num_batches) and size of 
+        progress bar while running.
+
+    defaults : (Optional) dict
+        A dictionary of default values for query. Should be of form:
+        {
+            "Address" : address_default,
+            "City" : city_default,
+            "County" : county_default,
+            "State" : state_default,
+            "Zip5" : postal_default,
+            "Country" : country_default
+        }
+        Will set country to United States by default.
+
+    keep_index : (Optional) bool
+        Whether to use the index of data for the geocoded result. If 
+        False, will use 
+        It is recommended to use the same index for the inputted data
+        to easily merge the data.
+        True by default.
+
+    Properties
+    ----------
+    street_encode : func
+        A function of 1 argument used to encode the street address. The 
+        default is street_encode(address) used to remove secondary 
+        descriptors from the street address. This is recommended 
+        because many addresses in OpenStreetMaps do not have
+        secondary descriptors attached and therefore will not be 
+        geocoded.
+
+    geocoder : str
+        Enter email as str to change geocoder user_agent.
+
+    email : str
+        Alias for geocoder
+
+    Methods
+    -------
+    run(k_batches = None)
+        Run num_batches batches of size batch_size through the geocoder.
+        If None will run all batches.
+
+    reset()
+        Method to reset progress and overwrite file at path.
+        WARNING: This method will overwrite file at path, do not use
+        unless you want to restart geocoding or if path is a path to 
+        existing file.
+
+    """
+
+    _cache_size = 4000
+    def __init__(self, data, email, path, batch_size = 3600,
+                 defaults = {"Country":"United States"}, 
+                 keep_index = True, index_name = ""):
+        """Constructor for bbd.geocoder.GeocodeLocations"""
+
+        #Creates geocoder with email passed as user_agent
+        self.geocoder = email #self.email is alias for self.geocoder
+        self.__real_geocoder = self._geocoder #for testing
+
+        self.defaults = defaults.copy()
```

Comment:
I see what you mean, but in the `else` statement it should be `self.defaults = defaults.copy()`, because defaults gets mutated to format all the keys for Nominatim.

Since I have to take a copy anyway I think it makes sense to define defaults explicitly in the `__init__(... defaults = {'Country':'United States'...)` as I don't have to worry about mutation b/c of base functionality.